src/bitops.c

```diff
@@ -254,6 +265,119 @@ long long redisPopCountAarch64(void *s, long count) {
+ATTRIBUTE_TARGET_AVX512_POPCOUNT
+long long redisPopCountAvx512(void *s, long count) {
+    long long bits = 0;
+    unsigned char *p = s;
+
+    /* Align to 64-byte boundary for optimal AVX512 performance */
+    while ((unsigned long)p & 63 && count) {
+        bits += bitsinbyte[*p++];
+        count--;
+    }
+
+    /* Process 64 bytes at a time using AVX512 */
+    while (count >= 64) {
+        __m512i data = _mm512_loadu_si512((__m512i*)p);
+        __m512i popcnt = _mm512_popcnt_epi64(data);
+
+        /* Sum all 8 64-bit popcount results */
+        bits += _mm512_reduce_add_epi64(popcnt);
+
+        p += 64;
+        count -= 64;
+
+        /* Prefetch next cache line */
+        redis_prefetch_read(p + 2048);
+    }
+
+    /* Handle remaining bytes with scalar popcount */
+    while (count >= 8) {
+        bits += __builtin_popcountll(*(uint64_t*)p);
+        p += 8;
+        count -= 8;
+    }
+
+    /* Handle final bytes */
+    while (count--) {
+        bits += bitsinbyte[*p++];
+    }
+
+    return bits;
+}
+#endif
+
+#ifdef HAVE_AVX2
+/* AVX2 optimized version of redisPopcount.
+ * This function requires AVX2 and POPCNT support. */
+ATTRIBUTE_TARGET_AVX2_POPCOUNT
+long long redisPopCountAvx2(void *s, long count) {
+    long long bits = 0;
+    unsigned char *p = s;
+
+    /* Align to 8-byte boundary for 64-bit operations */
+    while ((unsigned long)p & 7 && count) {
+        bits += bitsinbyte[*p++];
+        count--;
+    }
+
+    /* Use separate counters to avoid dependencies, similar to regular redisPopcount */
+    uint64_t cnt[4];
+    memset(cnt, 0, sizeof(cnt));
+
+    /* Process 32 bytes at a time using POPCNT on 64-bit chunks */
+    while (count >= 32) {
+        cnt[0] += __builtin_popcountll(*(uint64_t*)(p));
+        cnt[1] += __builtin_popcountll(*(uint64_t*)(p + 8));
+        cnt[2] += __builtin_popcountll(*(uint64_t*)(p + 16));
+        cnt[3] += __builtin_popcountll(*(uint64_t*)(p + 24));
+
+        p += 32;
+        count -= 32;
+
+        /* Prefetch next cache line */
+        redis_prefetch_read(p + 2048);
+    }
+
+    bits += cnt[0] + cnt[1] + cnt[2] + cnt[3];
+
+    /* Handle remaining bytes with scalar popcount */
+    while (count >= 8) {
+        bits += __builtin_popcountll(*(uint64_t*)p);
+        p += 8;
+        count -= 8;
+    }
+
+    /* Handle final bytes */
+    while (count--) {
+        bits += bitsinbyte[*p++];
+    }
+
+    return bits;
+}
+#endif
+
+/* Automatically select the best available popcount implementation */
+static inline long long redisPopcountAuto(const unsigned char *p, long count) {
+#ifdef HAVE_AVX512
+    if (BITOP_USE_AVX512) {
+        return redisPopCountAvx512((void*)p, count);
+    }
+#endif
+#ifdef HAVE_AVX2
+    if (BITOP_USE_AVX2) {
+        return redisPopCountAvx2((void*)p, count);
+    }
+#endif
+#ifdef HAVE_AARCH64_NEON
+    return redisPopCountAarch64((void*)p, count);
+#endif
+    return redisPopcount((void*)p, count);
```

Comment:
if HAVE_AARCH64_NEON is defined, there will be two returns.
i'm sure the compiler will give the complain.
```suggestion
#ifdef HAVE_AARCH64_NEON
    return redisPopCountAarch64((void*)p, count);
#else
    return redisPopcount((void*)p, count);
#endif
```